File: frontend/src/utils/pdfUtils.ts
// frontend/src/utils/pdfUtils.ts

import api from "../lib/api";

export interface VoucherPdfConfig {
  voucherType: string;
  voucherTitle: string;
  showItems?: boolean;
  showTaxDetails?: boolean;
  entityType?: "vendor" | "customer";
}
export interface VoucherPdfData {
  voucher_number: string;
  date: string;
  reference?: string;
  notes?: string;
  total_amount: number;
  items?: any[];
  // Entity information (vendor/customer)
  vendor?: {
    id: number;
    name: string;
    address?: string;
    contact_number?: string;
    email?: string;
    gstin?: string;
  };
  customer?: {
    id: number;
    name: string;
    address?: string;
    contact_number?: string;
    email?: string;
    gstin?: string;
  };
  // Voucher specific fields
  payment_method?: string;
  receipt_method?: string;
  payment_terms?: string;
  from_account?: string;
  to_account?: string;
  // Additional fields for different voucher types
  [key: string]: any;
}
/**
 * Generate PDF for any voucher type using standardized configuration
 */
export const generateVoucherPDF = async (
  voucherId: number,
  config: VoucherPdfConfig,
): Promise<void> => {
  try {
    // Check authorization before generating PDF
    const token = localStorage.getItem("token");
    if (!token) {
      alert("You must be logged in to generate PDFs");
      return;
    }
    // Call backend API for PDF generation
    const response = await api.post(
      `/pdf-generation/voucher/${config.voucherType}/${voucherId}/download`,
      {},
      { responseType: 'blob' }
    );
    // Handle the response blob
    const blob = response.data;
    const url = window.URL.createObjectURL(blob);

    // Improved filename extraction from Content-Disposition
    let filename = `${config.voucherTitle.replace(/\s+/g, '_')}_${voucherId}.pdf`; // Fallback filename
    const contentDisposition = response.headers['content-disposition'];
    console.log('Content-Disposition header:', contentDisposition);
    if (contentDisposition) {
      // Better regex to handle filename with or without quotes
      const filenameRegex = /filename\*?=['"]?(?:UTF-\d['"]*)?([^;\r\n"']*)['"]?;?/i;
      const matches = filenameRegex.exec(contentDisposition);
      if (matches && matches[1]) {
        filename = matches[1].replace(/['"]/g, '');
        console.log('Extracted filename:', filename);
      } else {
        console.warn('Failed to extract filename from header');
      }
    } else {
      console.warn('No Content-Disposition header found');
    }

    const a = document.createElement('a');
    a.href = url;
    a.download = filename;
    document.body.appendChild(a);
    a.click();
    document.body.removeChild(a);
    window.URL.revokeObjectURL(url);
  } catch (error) {
    console.error("Error generating PDF:", error);
    alert("Failed to generate PDF. Please try again.");
  }
};
/**
 * Voucher type configurations for PDF generation
 */
export const VOUCHER_PDF_CONFIGS: Record<string, VoucherPdfConfig> = {
  // Financial Vouchers
  "payment-voucher": {
    voucherType: "payment-voucher",
    voucherTitle: "PAYMENT VOUCHER",
    showItems: false,
    showTaxDetails: false,
    entityType: "vendor",
  },
  "receipt-voucher": {
    voucherType: "receipt-voucher",
    voucherTitle: "RECEIPT VOUCHER",
    showItems: false,
    showTaxDetails: false,
    entityType: "customer",
  },
  "journal-voucher": {
    voucherType: "journal-voucher",
    voucherTitle: "JOURNAL VOUCHER",
    showItems: false,
    showTaxDetails: false,
  },
  "contra-voucher": {
    voucherType: "contra-voucher",
    voucherTitle: "CONTRA VOUCHER",
    showItems: false,
    showTaxDetails: false,
  },
  // Purchase Vouchers
  "purchase-voucher": {
    voucherType: "purchase-voucher",
    voucherTitle: "PURCHASE VOUCHER / BILL",
    showItems: true,
    showTaxDetails: true,
    entityType: "vendor",
  },
  "purchase-order": {
    voucherType: "purchase-orders",
    voucherTitle: "PURCHASE ORDER",
    showItems: true,
    showTaxDetails: true,
    entityType: "vendor",
  },
  grn: {
    voucherType: "grn",
    voucherTitle: "GOODS RECEIPT NOTE",
    showItems: true,
    showTaxDetails: false,
    entityType: "vendor",
  },
  "purchase-return": {
    voucherType: "purchase-return",
    voucherTitle: "PURCHASE RETURN",
    showItems: true,
    showTaxDetails: true,
    entityType: "vendor",
  },
  // Sales Vouchers
  "sales-voucher": {
    voucherType: "sales-voucher",
    voucherTitle: "SALES INVOICE",
    showItems: true,
    showTaxDetails: true,
    entityType: "customer",
  },
  quotation: {
    voucherType: "quotation",
    voucherTitle: "QUOTATION",
    showItems: true,
    showTaxDetails: true,
    entityType: "customer",
  },
  "proforma-invoice": {
    voucherType: "proforma-invoice",
    voucherTitle: "PROFORMA INVOICE",
    showItems: true,
    showTaxDetails: true,
    entityType: "customer",
  },
  "sales-order": {
    voucherType: "sales-order",
    voucherTitle: "SALES ORDER",
    showItems: true,
    showTaxDetails: true,
    entityType: "customer",
  },
  "delivery-challan": {
    voucherType: "delivery-challan",
    voucherTitle: "DELIVERY CHALLAN",
    showItems: true,
    showTaxDetails: false,
    entityType: "customer",
  },
  "sales-return": {
    voucherType: "sales-return",
    voucherTitle: "SALES RETURN",
    showItems: true,
    showTaxDetails: true,
    entityType: "customer",
  },
  "credit-note": {
    voucherType: "credit-note",
    voucherTitle: "CREDIT NOTE",
    showItems: true,
    showTaxDetails: true,
    entityType: "customer",
  },
  "debit-note": {
    voucherType: "debit-note",
    voucherTitle: "DEBIT NOTE",
    showItems: true,
    showTaxDetails: true,
    entityType: "vendor",
  },
  "non-sales-credit-note": {
    voucherType: "non-sales-credit-note",
    voucherTitle: "NON-SALES CREDIT NOTE",
    showItems: true,
    showTaxDetails: true,
    entityType: "customer",
  },
  // Manufacturing Vouchers
  "job-card": {
    voucherType: "job-card",
    voucherTitle: "JOB CARD",
    showItems: true,
    showTaxDetails: true,
    entityType: "vendor",
  },
  "production-order": {
    voucherType: "production-order",
    voucherTitle: "PRODUCTION ORDER",
    showItems: true,
    showTaxDetails: false,
  },
  "work-order": {
    voucherType: "work-order",
    voucherTitle: "WORK ORDER",
    showItems: true,
    showTaxDetails: false,
  },
  "material-receipt": {
    voucherType: "material-receipt",
    voucherTitle: "MATERIAL RECEIPT",
    showItems: true,
    showTaxDetails: false,
  },
  "material-requisition": {
    voucherType: "material-requisition",
    voucherTitle: "MATERIAL REQUISITION",
    showItems: true,
    showTaxDetails: false,
  },
  "finished-good-receipt": {
    voucherType: "finished-good-receipt",
    voucherTitle: "FINISHED GOODS RECEIPT",
    showItems: true,
    showTaxDetails: false,
  },
  "manufacturing-journal": {
    voucherType: "manufacturing-journal",
    voucherTitle: "MANUFACTURING JOURNAL",
    showItems: false,
    showTaxDetails: false,
  },
  "stock-journal": {
    voucherType: "stock-journal",
    voucherTitle: "STOCK JOURNAL",
    showItems: true,
    showTaxDetails: false,
  },
};
/**
 * Get PDF configuration for a voucher type
 */
export const getVoucherPdfConfig = (voucherType: string): VoucherPdfConfig => {
  const config = VOUCHER_PDF_CONFIGS[voucherType];
  if (!config) {
    console.warn(`No PDF configuration found for voucher type: {voucherType}`);
    return {
      voucherType,
      voucherTitle: voucherType.toUpperCase().replace(/-/g, " "),
      showItems: false,
      showTaxDetails: false,
    };
  }
  return config;
};
/**
 * Standalone PDF generation function for individual vouchers
 * Can be used in any voucher component without requiring useVoucherPage hook
 */
export const generateStandalonePDF = async (
  voucherId: number,
  voucherType: string,
): Promise<void> => {
  try {
    console.log(
      "[PDF] Generating standalone PDF for:",
      voucherType,
      voucherId,
    );
    // Check authorization
    const token = localStorage.getItem("token");
    if (!token) {
      alert("You must be logged in to generate PDFs");
      return;
    }
    // Get PDF configuration
    const pdfConfig = getVoucherPdfConfig(voucherType);
    // Generate PDF via backend
    await generateVoucherPDF(voucherId, pdfConfig);
  } catch (error) {
    console.error("[PDF] Error generating standalone PDF:", error);
    alert("Failed to generate PDF. Please try again.");
  }
};

File: app/utils/pdf_utils.py
# app/utils/pdf_utils.py

"""
Unified PDF generation utility for all voucher types
"""

from typing import Dict

class VoucherPdfConfig:
    def __init__(self, voucher_type: str, voucher_title: str, show_items: bool = False, show_tax_details: bool = False, entity_type: str = None):
        self.voucher_type = voucher_type
        self.voucher_title = voucher_title
        self.show_items = show_items
        self.show_tax_details = show_tax_details
        self.entity_type = entity_type

VOUCHER_PDF_CONFIGS: Dict[str, VoucherPdfConfig] = {
    # Financial Vouchers
    "payment-voucher": VoucherPdfConfig(
        voucher_type="payment-vouchers",
        voucher_title="PAYMENT VOUCHER",
        show_items=False,
        show_tax_details=False,
        entity_type="vendor",
    ),
    "receipt-voucher": VoucherPdfConfig(
        voucher_type="receipt-vouchers",
        voucher_title="RECEIPT VOUCHER",
        show_items=False,
        show_tax_details=False,
        entity_type="customer",
    ),
    "journal-voucher": VoucherPdfConfig(
        voucher_type="journal-vouchers",
        voucher_title="JOURNAL VOUCHER",
        show_items=False,
        show_tax_details=False,
    ),
    "contra-voucher": VoucherPdfConfig(
        voucher_type="contra-vouchers",
        voucher_title="CONTRA VOUCHER",
        show_items=False,
        show_tax_details=False,
    ),
    # Purchase Vouchers
    "purchase-voucher": VoucherPdfConfig(
        voucher_type="purchase-vouchers",
        voucher_title="PURCHASE VOUCHER / BILL",
        show_items=True,
        show_tax_details=True,
        entity_type="vendor",
    ),
    "purchase-order": VoucherPdfConfig(
        voucher_type="purchase-orders",
        voucher_title="PURCHASE ORDER",
        show_items=True,
        show_tax_details=True,
        entity_type="vendor",
    ),
    "grn": VoucherPdfConfig(
        voucher_type="grns",
        voucher_title="GOODS RECEIPT NOTE",
        show_items=True,
        show_tax_details=False,
        entity_type="vendor",
    ),
    "purchase-return": VoucherPdfConfig(
        voucher_type="purchase-returns",
        voucher_title="PURCHASE RETURN",
        show_items=True,
        show_tax_details=True,
        entity_type="vendor",
    ),
    # Sales Vouchers
    "sales-voucher": VoucherPdfConfig(
        voucher_type="sales-vouchers",
        voucher_title="SALES INVOICE",
        show_items=True,
        show_tax_details=True,
        entity_type="customer",
    ),
    "quotation": VoucherPdfConfig(
        voucher_type="quotations",
        voucher_title="QUOTATION",
        show_items=True,
        show_tax_details=True,
        entity_type="customer",
    ),
    "proforma-invoice": VoucherPdfConfig(
        voucher_type="proforma-invoices",
        voucher_title="PROFORMA INVOICE",
        show_items=True,
        show_tax_details=True,
        entity_type="customer",
    ),
    "sales-order": VoucherPdfConfig(
        voucher_type="sales-orders",
        voucher_title="SALES ORDER",
        show_items=True,
        show_tax_details=True,
        entity_type="customer",
    ),
    "delivery-challan": VoucherPdfConfig(
        voucher_type="delivery-challans",
        voucher_title="DELIVERY CHALLAN",
        show_items=True,
        show_tax_details=False,
        entity_type="customer",
    ),
    "sales-return": VoucherPdfConfig(
        voucher_type="sales-returns",
        voucher_title="SALES RETURN",
        show_items=True,
        show_tax_details=True,
        entity_type="customer",
    ),
    "credit-note": VoucherPdfConfig(
        voucher_type="credit-notes",
        voucher_title="CREDIT NOTE",
        show_items=True,
        show_tax_details=True,
        entity_type="customer",
    ),
    "debit-note": VoucherPdfConfig(
        voucher_type="debit-notes",
        voucher_title="DEBIT NOTE",
        show_items=True,
        show_tax_details=True,
        entity_type="vendor",
    ),
    "non-sales-credit-note": VoucherPdfConfig(
        voucher_type="non-sales-credit-notes",
        voucher_title="NON-SALES CREDIT NOTE",
        show_items=True,
        show_tax_details=True,
        entity_type="customer",
    ),
    # Manufacturing Vouchers
    "job-card": VoucherPdfConfig(
        voucher_type="job-cards",
        voucher_title="JOB CARD",
        show_items=True,
        show_tax_details=True,
        entity_type="vendor",
    ),
    "production-order": VoucherPdfConfig(
        voucher_type="production-orders",
        voucher_title="PRODUCTION ORDER",
        show_items=True,
        show_tax_details=False,
    ),
    "work-order": VoucherPdfConfig(
        voucher_type="work-orders",
        voucher_title="WORK ORDER",
        show_items=True,
        show_tax_details=False,
    ),
    "material-receipt": VoucherPdfConfig(
        voucher_type="material-receipts",
        voucher_title="MATERIAL RECEIPT",
        show_items=True,
        show_tax_details=False,
    ),
    "material-requisition": VoucherPdfConfig(
        voucher_type="material-requisitions",
        voucher_title="MATERIAL REQUISITION",
        show_items=True,
        show_tax_details=False,
    ),
    "finished-good-receipt": VoucherPdfConfig(
        voucher_type="finished-good-receipts",
        voucher_title="FINISHED GOODS RECEIPT",
        show_items=True,
        show_tax_details=False,
    ),
    "manufacturing-journal": VoucherPdfConfig(
        voucher_type="manufacturing-journals",
        voucher_title="MANUFACTURING JOURNAL",
        show_items=False,
        show_tax_details=False,
    ),
    "stock-journal": VoucherPdfConfig(
        voucher_type="stock-journals",
        voucher_title="STOCK JOURNAL",
        show_items=True,
        show_tax_details=False,
    ),
}

def get_voucher_pdf_config(voucher_type: str) -> VoucherPdfConfig:
    config = VOUCHER_PDF_CONFIGS.get(voucher_type)
    if not config:
        logger.warning(f"No PDF configuration found for voucher type: {voucher_type}")
        return VoucherPdfConfig(
            voucher_type=voucher_type,
            voucher_title=voucher_type.upper().replace("-", " "),
            show_items=False,
            show_tax_details=False,
        )
    return config

def generate_standalone_pdf(voucher_id: int, voucher_type: str) -> None:
    logger.info(f"[PDF] Generating standalone PDF for: {voucher_type} {voucher_id}")
    # Check authorization
    token = "dummy_token"  # Replace with actual token retrieval
    if not token:
        raise ValueError("You must be logged in to generate PDFs")
    # Get PDF configuration
    pdf_config = get_voucher_pdf_config(voucher_type)
    # Generate PDF via backend
    generate_voucher_pdf(voucher_id, pdf_config)  # Assuming this function exists